Applications/ClientWebPortal/Source/webapp/src/utils/chainable-models/indexed-model.js

```diff
@@ -0,0 +1,76 @@
+import ArrayModel from './array-model';
+import HashMap from 'hashmap';
+import ArrayStringKeyGenerator from 'utils/array-string-key-generator';
+
+export default class {
+  constructor(indices, columnNames, keyGenerator = new ArrayStringKeyGenerator()) {
+    // validate indices
+    let columnLength = columnNames.length;
+    if (indices == null || indices.length > columnNames) {
+      let errorMessage = 'Indices has to be an array of column numbers';
+      throw new TypeError(errorMessage);
+    } else {
+      for (let i=0; i<indices.length; i++) {
+        if (indices[i] < 0 || indices[i] >= columnNames.length) {
+          let errorMessage = 'Index ' + indices[i] + ' is out of bounds';
+          throw new RangeError(errorMessage);
+        }
+      }
+    }
+
+    this.indices = indices;
+    this.arrayModel = new ArrayModel(columnNames);
+    this.keyGenerator = keyGenerator;
+    this.indicesToRow = new HashMap();
+  }
+
+  getRowCount() {
+    return this.arrayModel.getRowCount();
+  }
+
+  getColumnCount() {
+    return this.arrayModel.getColumnCount();
+  }
+
+  getColumnName(columnIndex) {
+    return this.arrayModel.getColumnName(columnIndex);
+  }
+
+  getValueAt(x, y) {
+    return this.arrayModel.getValueAt(x, y);
+  }
+
+  update(values) {
+    let indexValues = [];
+    for (let i=0; i<this.indices.length; i++) {
+      indexValues.push(values[i]);
+    }
+    let key = this.keyGenerator.get(indexValues);
+
+    if (!this.indicesToRow.has(key)) {
+      let rowNumber = this.arrayModel.addRow(values);
```

Comment:
This line introduces a data inconsistency.  A signal will be emitted as soon as the addRow method is called meaning all signal handlers will see an incomplete update.

To avoid this inconsistency you will need to ensure that adding the row is the very last operation you perform.  To ensure this, you will rearrange your code as follows:

  let rowNumber = this.arrayModel.getRowCount();
  this.indicesToRow.set(key, rowNumber);
  this.arrayModel.addRow(values);

This ensures that all signal handlers get a proper/consistent view of the Model.

Reply:
Will do.

Reply:
You would need to do 
let rowNumber = this.arrayModel.getRowCount();
this.indicesToRow.set(key, rowNumber + 1);
this.arrayModel.addRow(values);

Reply:
No the original code snippet I provided is correct.

Consider the case of an empty table.

let rowNumber = this.arrayModel.getRowCount();
  // This implies that rowNumber == 0;
this.indicesToRow.set(key, rowNumber + 1);
  // This implies that the index maps to row 1, which is incorrect, the correct row is 0.

Reply:
Oh shoot. Right. Row count returns the "length" of the array. Not the last index. For some reason I thought that. My bad.

Reply:
Even if I tried + 1, unit test would have caught it.